File: benchmarks/c/numeric_loop.c
#include <stdio.h>

long collatz_steps(long start) {
    long n = start;
    long steps = 0;
    while (n != 1) {
        if (n % 2 == 0)
            n = n / 2;
        else
            n = n * 3 + 1;
        steps++;
    }
    return steps;
}

int main(void) {
    long total_steps = 0;
    for (long i = 1; i < 1000000; i++) {
        total_steps += collatz_steps(i);
    }
    printf("Total Collatz steps: %ld\n", total_steps);
    return 0;
}

File: benchmarks/ocaml/numeric_loop.ml
(* Pure numeric loop benchmark: Collatz sequence. *)

let collatz_steps start =
  let n = ref start in
  let steps = ref 0 in
  while !n <> 1 do
    if !n mod 2 = 0 then n := !n / 2 else n := (!n * 3) + 1;
    incr steps
  done;
  !steps

let () =
  let total_steps = ref 0 in
  for i = 1 to 999_999 do
    total_steps := !total_steps + collatz_steps i
  done;
  Printf.printf "Total Collatz steps: %d\n" !total_steps
